doc/neps/nep-0043-extensible-ufuncs.rst

```diff
@@ -0,0 +1,631 @@
+
+* Similar to casting, also ufuncs may need to find the correct output DType
+  or indicate that a loop is only capable of handling certain instances of
+  the involved DTypes (e.g. only native byte order).  This is handled by
+  an ``resolve_descriptors`` function (similar to ``adjust_descriptors``
+  of ``CastingImpl``)::
+
+      int resolve_descriptors(
+              PyUFuncObject *ufunc, PyArray_DTypeMeta *DTypes[nargs],
+              PyArray_Descr *in_dtypes[nargs], PyArray_Descr *out_dtypes[nargs]);
+
+  setting the desired output descriptors, and possibly modifying the input
+  ones to conform with the inner-loops requirements.
+
+* Define a new structure to be passed in the inner-loop, which can be
+  partially modified by the setup/teardown as well::
+  
+      typedef struct {
+          PyUFuncObject *ufunc,
+          /* could add information about __call__, reduce, etc. here */
+          // method
+          /* The exact operand dtypes of the inner-loop */
+          PyArray_Descr const *dtypes;
+          /* 
+           * The Python threadstate (if the GIL was released) or NULL.
+           * Will always be NULL, if `needs_api` was indicated.
+           */
+          PyThreadState const *tstate;
+          /* Per-loop user data, equivalent to the current void* */
+          void const *user_loop_data;
+          /* User data potentially filled by the setup function. */
+          void *user_data;
+      } PyArray_UFuncData
+  
+  This structure may be appended to include additional information in future
+  versions of NumPy.
+
+* The *optional* setup function::
+
+      innerloop *
+      setup(PyArray_UFuncData *data, int *needs_api, npy_intp *fixed_strides);
+  
+  Is passed the above struct and may modify (only) the ``user_data`` slot
+  and potentially further slots in the future.  The function returns
+  the inner-loop or ``NULL`` to indicate an error.
+
+* The inner-loop function::
+
+    int inner_loop(..., PyArray_UFuncData *data);
+
+  Will have the identical signature to current inner-loops with the difference
+  of the additional return value and passing ``PyArray_UfuncData *`` instead
+  of the previous ``void *`` representing ``user_loop_data``.
+  The inner-loop shall return 0 to indicate continued (successful) execution.
+  Any non-zero value will terminate the iteration.  The inner-loop shall
+  return a *negative* value (e.g. -1) with a Python exception set.  
+
+* The *optional* teardown function::
+
+     int teardown(int return_value, PyArray_UFuncData *data);
+
+  Teardown function, may return ``-1`` if an error is set.  If an error
+  was already set previously it will be stored. Any previous error
+  will be cleared before calling ``teardown``.
+  The return value of the inner-loop is passed in to allow simple error
+  flagging (e.g. warnings) if desired.
+
+* A flag to ask NumPy to perform floating point error checking (after custom
+  setup and before user teardown).
```

Comment:
Many integer loops or loops working with Python objects only, do not really have to. And apparently floating point error checking can be a bit slow on some hardware, so I think we can create such an option.  Although, I guess it would be fair to add that later as well.